packages/main/src/YearRangePicker.ts

```diff
@@ -0,0 +1,510 @@
+	 * @default []
+	 */
+	@property({ type: Array })
+	selectedDates: Array<number> = [];
+	/**
+	 * Defines the type of selection used in the parent calendar component.
+	 * Note that CalendarSelection is not available for the YearRangePicker, this property is only to visualize the selected dates.
+	 * @default "Single"
+	 * @since 2.2.0
+	 */
+	@property()
+	_selectionMode: `${CalendarSelectionMode}` = "Single";
+
+	/**
+	 * When selectionMode="Range" and the first year in the range is selected, this is the currently hovered or focused year.
+	 *
+	 * @private
+	 */
+	@property({ type: Number })
+	_secondTimestamp?: number;
+
+	@property({ type: Array })
+	_yearRanges: YearRanges = [];
+
+	@property({ type: Boolean, noAttribute: true })
+	_hidden = false;
+
+	@property({ noAttribute: true })
+	_currentYearRange?: CalendarYearRangeT;
+
+	_gridStartYear?: number;
+
+	@i18n("@ui5/webcomponents")
+	static i18nBundle: I18nBundle;
+
+	get roleDescription() {
+		return YearRangePicker.i18nBundle.getText(YEAR_RANGE_PICKER_DESCRIPTION);
+	}
+
+	onBeforeRendering() {
+		if (this._hidden) {
+			return;
+		}
+
+		this._gridStartYear = this._calculateGridStartYear();
+		this._buildYears();
+	}
+
+	_shouldShowOneColumn() {
+		const locale = getLocale() as unknown as LocaleT;
+		const language = locale.getLanguage();
+		const longLanguage = language === "zh" || language === "ja" || language === "ko" || language === "bg" || language === "mk" || language === "ru";
+
+		return longLanguage && this.hasSecondaryCalendarType;
+	}
+
+	_getPageSize() {
+		return this._shouldShowOneColumn() ? 6 : 8;
+	}
+
+	_getRowSize() {
+		return this._shouldShowOneColumn() ? 1 : 2;
+	}
+
+	_getInitialFocusedIndex() {
+		return 2;
+	}
+
+	_getRangeSize() {
+		return this.hasSecondaryCalendarType ? 8 : 20;
+	}
+
+	_calculateGridStartYear() {
+		const rangeSize = this._getRangeSize();
+		const pageSize = this._getPageSize();
+		const pageSizeInYears = rangeSize * pageSize;
+		const yearsOffset = rangeSize * this._getInitialFocusedIndex();
+		const currentStartYear = this._currentYearRange?.startYear ? this._currentYearRange?.startYear : this._calendarDate.getYear();
+
+		// On first load, current range should be the 3rd item in the grid
+		let gridStartYear = this._gridStartYear ? this._gridStartYear : currentStartYear - yearsOffset;
+
+		// If page navigation occured, update the current range start year
+		gridStartYear += Math.floor((currentStartYear - gridStartYear) / pageSizeInYears) * pageSizeInYears;
+
+		// Normalize grid start year to be between the min and absolute max year
+		const minYear = this._minDate.getYear();
+		if (currentStartYear - rangeSize < minYear) {
+			gridStartYear = minYear;
+		}
+
+		const absoluteMaxYear = getMaxCalendarDate(this._primaryCalendarType).getYear();
+		if (currentStartYear + pageSizeInYears > absoluteMaxYear) {
+			gridStartYear = absoluteMaxYear - pageSizeInYears + 1;
+		}
+
+		return gridStartYear;
+	}
+
+	_buildYears() {
```

Comment:
build year ranges?
and can we split this method as it is close to 100 lines of code?